### PR TITLE
GH1089: Add additional context information to Teardown

### PR DESCRIPTION
**Pull request by gep13**
Relates to #1089

### CODE REVIEW

**Comment by gep13** [comment]
@cake-build/team Apologies for creating a branch on this repository.  I was already completed the work locally before I noticed :cry:

**Comment by gep13** [comment]
@cake-build/team what are your thoughts on this?

This would allow the following:

```
Teardown((context, build) =>
{
    if(build.Successful)
    {
        Information("Build successful");
    }
    else
    {
        Information("Build failed");
    }
});
```

Tested locally using Cake build as test, and seems to work as I would expect it to:

image

I basically threw an exception in the package task for testing it out.

**Comment by patriksvensson** [comment]
@gep13 If `ITearDownContext` would inherit `ICakeContext` then we would be backwards compatible without adding a new overload for `Teardown`.

**Comment by gep13** [comment]
@patriksvensson I did think about that, but then I thought it would be "different" from others in the same file, so thought consistency was better.  Unless we also go through and change the others as well?

**Comment by gep13** [comment]
@patriksvensson or, should we do this one in the "new" way, and then create another issue to go back and change the existing ones?

**Comment by patriksvensson** [comment]
Maybe a good idea to change them all. What do you think?

**Comment by gep13** [comment]
I can try to take a stab at that :-)

**Comment by devlead** [comment]
An idea, maybe
https://github.com/cake-build/cake/blob/1f2c8b9d677cdc55df33c6d9c3c036fa056c3c45/src/Cake.Core/CakeEngine.cs#L149-L157
```csharp
            catch
            {
                exceptionWasThrown = true;
                throw;
            }
            finally
            {
                PerformTeardown(strategy, context, exceptionWasThrown);
            }
```
should be something like
```csharp
            catch(Exception ex)
            {
                exceptionWasThrown = true;
                thrownException = ex;
                throw;
            }
            finally
            {
                PerformTeardown(strategy, context, exceptionWasThrown, ex);
            }
```
so we could report exception back to on teardown context.

Thoughts?

**Comment by gep13** [comment]
>@devlead said...
Should be added to ITeardownContext interface too right?

Yip, last night's commit was an interim one :-D

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 1]
```diff
@@ -683,7 +683,7 @@ public void Should_Run_Task_Setup_Before_Each_Task()
                 var result = new List<string>();
                 var fixture = new CakeEngineFixture();
                 var engine = fixture.CreateEngine();
-                engine.RegisterTaskSetupAction((cc, sc) => result.Add("TASK_SETUP:" + sc.Task.Name));
+                engine.RegisterTaskSetupAction(context => result.Add("TASK_SETUP:" + ((ITaskSetupContext)context).Task.Name));
```
Comment (patriksvensson) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: Should we have to cast the context here?

File: src/Cake.Core.Tests/Unit/TaskSetupContextTests.cs [hunk 2]
```diff
@@ -13,8 +14,11 @@ public sealed class TheConstructor
             [Fact]
             public void Should_Throw_If_Task_Is_Null()
```
Comment (patriksvensson) [inline src/Cake.Core.Tests/Unit/TaskSetupContextTests.cs]: Maybe we should add a test for checking that it throws if the context is null as well.

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs (hunk 1 above)
Comment (gep13) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: No, not any more. Originally I did, then I changed it based on your feedback. Will correct.

File: src/Cake.Core/TeardownContext.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TeardownContext"/> class.
+        /// </summary>
+        /// <param name="cakeContext">The Cake Context</param>
+        /// <param name="successful">If set to <c>true</c>, the build was not successful.</param>
+        /// <param name="throwException">The exception that was thrown by the target.</param>
+        public TeardownContext(ICakeContext cakeContext, bool successful, Exception throwException)
+            : base(cakeContext)
+        {
+            _successful = successful;
```
Comment (patriksvensson) [inline src/Cake.Core/TeardownContext.cs]: Would a successful build be the same as a build with no exceptions? In that case, maybe we can just check `thrownException` for null instead of providing success. Or do we want this to be explicit? Either one is fine by me, just asking the question loud :smile:

File: src/Cake.Core.Tests/Unit/TaskTeardownContextTests.cs [hunk 4]
```diff
@@ -10,10 +12,23 @@ public sealed class TaskTeardownContextTests
         public sealed class TheConstructor
         {
             [Fact]
-            public void Should_Throw_If_Task_Is_Null()
+            public void Should_Throw_If_Context_Is_Null()
             {
                 // Given, When
-                var result = Record.Exception(() => new TaskSetupContext(null));
+                var result = Record.Exception(() => new TaskSetupContext(null, null));
```
Comment (patriksvensson) [inline src/Cake.Core.Tests/Unit/TaskTeardownContextTests.cs]: Should provide a task here.

File: src/Cake.Core.Tests/Unit/TeardownContextTests.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+{
+    public sealed class TeardownContextTests
+    {
+        public sealed class TheConstructor
+        {
+            [Fact]
+            public void Should_Throw_If_Context_Is_Null()
+            {
+                // Given, When
+                var result = Record.Exception(() => new TeardownContext(null, null));
```
Comment (patriksvensson) [inline src/Cake.Core.Tests/Unit/TeardownContextTests.cs]: Should provide a task here.

File: src/Cake.Core/CakeContextAdapter.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    /// </summary>
+    public abstract class CakeContextAdapter
+    {
+        private readonly ICakeContext _cakeContext;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CakeContextAdapter"/> class.
+        /// </summary>
+        /// <param name="cakeContext">The Cake Context</param>
+        protected CakeContextAdapter(ICakeContext cakeContext)
```
Comment (patriksvensson) [inline src/Cake.Core/CakeContextAdapter.cs]: Can we rename this to `context`?

File: src/Cake.Core/TaskSetupContext.cs [hunk 7]
```diff
@@ -8,15 +8,17 @@ namespace Cake.Core
...
     {
         private readonly ICakeTaskInfo _task;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="TaskSetupContext"/> class.
         /// </summary>
+        /// <param name="cakeContext">The Cake Context.</param>
         /// <param name="task">The task.</param>
-        public TaskSetupContext(ICakeTaskInfo task)
+        public TaskSetupContext(ICakeContext cakeContext, ICakeTaskInfo task)
```
Comment (patriksvensson) [inline src/Cake.Core/TaskSetupContext.cs]: Can we rename this to `context`?

File: src/Cake.Core/TaskTeardownContext.cs [hunk 8]
```diff
@@ -17,10 +17,12 @@ public sealed class TaskTeardownContext : ITaskTeardownContext
         /// <summary>
         /// Initializes a new instance of the <see cref="TaskTeardownContext"/> class.
         /// </summary>
+        /// <param name="cakeContext">The Cake Context.</param>
         /// <param name="task">The task.</param>
         /// <param name="duration">The duration of the task.</param>
         /// <param name="skipped">if set to <c>true</c>, the task was not executed.</param>
-        public TaskTeardownContext(ICakeTaskInfo task, TimeSpan duration, bool skipped)
+        public TaskTeardownContext(ICakeContext cakeContext, ICakeTaskInfo task, TimeSpan duration, bool skipped)
```
Comment (patriksvensson) [inline src/Cake.Core/TaskTeardownContext.cs]: Can we rename this to `context`?

File: src/Cake.Core/TeardownContext.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    public sealed class TeardownContext : CakeContextAdapter, ITeardownContext
+    {
+        private readonly Exception _thrownException;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TeardownContext"/> class.
+        /// </summary>
+        /// <param name="cakeContext">The Cake Context</param>
+        /// <param name="throwException">The exception that was thrown by the target.</param>
+        public TeardownContext(ICakeContext cakeContext, Exception throwException)
```
Comment (patriksvensson) [inline src/Cake.Core/TeardownContext.cs]: Can we rename this to `context`?

**Comment by gep13** [comment]
@patriksvensson I think I have addressed all your comments :smile:

**Comment by patriksvensson** [comment]
@gep13 Looks good to me! Do you feel ready with this? Should I merge it?
